apps/sms/js/recipients.js

```diff
@@ -290,45 +308,21 @@
       }
     });
 
-    // Don't bother rejecting duplicates, always add every
-    // entry to the recipients list. For reference, see:
-    // https://bugzilla.mozilla.org/show_bug.cgi?id=880628
-    list.push(added = new Recipient(entry));
-
-    // XXX: Workaround for cleaning search result while duplicate
-    //      Dispatch add event no matter duplicate or not.
-    //
-    //      Send a "clone" of the added recipient, this protects
-    //      the actual Recipient object from being written to
-    //      directly.
-    this.emit('add', list.length, added.clone());
+    var recipient = new Recipient(entry);
 
-    // Render the view
-    this.render();
+    this.validator.validate(recipient).then((validatedRecipient) => {
+      var list = data.get(this);
 
-    return this;
-  };
+      // Don't bother rejecting duplicates, always add every
+      // entry to the recipients list. For reference, see:
+      // https://bugzilla.mozilla.org/show_bug.cgi?id=880628
+      list.push(Recipient.cast(validatedRecipient));
 
-  /**
-   * update Update/replace a recipient at a specific index in the list.
-   *
-   * @param  {Number} recipOrIndex Recipient or Index to update.
-   * @param  {Object} entry { name: '', number: '' }.
-   *
-   * @return {Recipients} return the recipients list.
-   */
-  Recipients.prototype.update = function(recipOrIndex, entry) {
-    var list = data.get(this);
-    var index = typeof recipOrIndex === 'number' ?
-      recipOrIndex : list.indexOf(recipOrIndex);
-
-    if (index > -1) {
-      // Use the normalization of new Recipient() to
-      // correct any missing, but required fields.
-      list[index].set(new Recipient(entry));
-    }
+      this.render();
+      this.emit('add');
```

Comment:
Not for this patch, but I'm wondering if ```EventDispatcher``` fulfills all Recipients needs so that we can get rid of custom ```emit/on/off```?